xla/service/collective_quantizer.cc

```diff
@@ -50,14 +60,121 @@ auto ConvertToNarrowerType() {
+  } else if (instr->opcode() == HloOpcode::kMultiply) {
+    for (HloInstruction* operand : instr->unique_operands()) {
+      auto binary_subgraph = FindDequantizationSubgraphRecursive(
+          operand, visited_instrs, subgraph);
+      if (!binary_subgraph.empty()) {
+        return binary_subgraph;
+      }
+    }
+  }
+  return {};
+}
+
+// Returns true iff instr describes a dequantization, i.e. a multiplication or
+// division by a broadcasted scalar operating on a type conversion, or a plain
+// type conversion to a wider type. Unary bitcast, copy, reshape or slice ops
+// may follow the dequantization or type conversion.
+std::optional<ConversionSubgraph> IsSupportedDequantization(
+    HloInstruction* instr) {
+  ConversionSubgraph subgraph;
+  absl::flat_hash_set<int> visited_instrs;
+  std::vector<HloInstruction*> candidate_subgraph =
+      FindDequantizationSubgraphRecursive(instr, visited_instrs,
+                                          std::vector<HloInstruction*>{});
+  std::reverse(candidate_subgraph.begin(), candidate_subgraph.end());
+
+  // In the dequantization case, the type conversion is followed by a
+  // multiplication or division by a broadcasted scalar.
+  if (candidate_subgraph.size() > 1 &&
+      (Match(
+           candidate_subgraph[1],
+           m::MultiplyAnyOrder(&subgraph.binary, m::Convert(&subgraph.convert),
+                               ScalarBroadcast(&subgraph.scale_bcast))) ||
+       Match(candidate_subgraph[1],
+             m::Divide(&subgraph.binary, m::Convert(&subgraph.convert),
+                       ScalarBroadcast(&subgraph.scale_bcast))))) {
+    subgraph.unaries = {candidate_subgraph.begin() + 2,
+                        candidate_subgraph.end()};
+  } else if (candidate_subgraph.size() > 0 &&
+             Match(candidate_subgraph[0], m::Convert(&subgraph.convert))) {
+    subgraph.unaries = {candidate_subgraph.begin() + 1,
+                        candidate_subgraph.end()};
+  } else {
+    VLOG(5) << "Did not find type conversion or dequantization pattern.";
+    return std::nullopt;
+  }
+
+  // The collected unary ops between dequantization/type conversion and
+  // collective may only include bitcast, copy, reshape and slice instructions.
+  for (HloInstruction* unary : subgraph.unaries) {
+    if (!Match(unary, m::AnyOf<HloInstruction>(m::Bitcast(), m::Copy(),
+                                               m::Reshape(), m::Slice()))) {
+      VLOG(5) << "Unexpected instruction in unary ops.";
+      return std::nullopt;
+    }
+  }
+  return std::make_optional<ConversionSubgraph>(subgraph);
+}
+
 // Returns true iff instr describes a quantization, i.e. a multiplication or
 // division by a broadcasted scalar followed by a clamp and a type conversion,
 // or a plain type conversion to a narrower type. Unary bitcast, copy, reshape
 // or slice ops with one user may precede the quantization or type conversion.
```

Comment:
In the comment, clarify that `instr` refers to the first, top-most such op in a quantization pattern. As contrasted with IsSupportedDequantization, where `instr` refers to the last, bottom-most op in the dequantization pattern.